src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities;
+
+internal static class AsyncLazy
+{
+    public static AsyncLazy<T, TArg> Create<T, TArg>(Func<TArg, CancellationToken, Task<T>> asynchronousComputeFunction, TArg arg)
+        => new AsyncLazy<T, TArg>(asynchronousComputeFunction, arg);
+
+    public static AsyncLazy<T, TArg> Create<T, TArg>(Func<TArg, CancellationToken, T> synchronousComputeFunction, TArg arg)
+        => new AsyncLazy<T, TArg>((arg, cancellationToken) => Task.FromResult(synchronousComputeFunction(arg, cancellationToken)), synchronousComputeFunction, arg);
```

Comment:
💡 This allocation somewhat defeats the purpose of `TArg`. It can be avoided by making `AsyncLazy<T, TArg>` support a null `asynchronousComputeFunction` in the specific case where `synchronousComputeFunction` is provided. It can also be avoided by changing the return type as mentioned in my other comment.

Reply:
Changed similar to suggestion